Simplify adapter construction; defer connect until first use

matthewd opened the pull request:
Building on from #44576, we no longer need to connect to the database in order to construct an adapter instance: we instead supply just the config hash, and it will connect on demand, in the same way it already knows to handle later reconnections.

This cleans up a weird API overlap, where we previously needed to know how to connect in two different places.

matthewd committed:
Simplify adapter construction; defer connect until first use

eileencodes commented on activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:
```diff
@@ -89,34 +89,51 @@ def self.quoted_table_names # :nodoc:
+          @logger = ActiveRecord::Base.logger
+
+          if deprecated_logger || deprecated_connection_options || deprecated_config
+            raise ArgumentError, "new API accepts just one config hash"
```
This is a confusing error because if I passed a config hash and a logger I wouldn't know what this error was indicating I should change. I know this is a WIP but wanted to note I think each deprecated option should have it's own error to be clearer about the new API.

eileencodes commented on activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:
```diff
@@ -89,34 +89,51 @@ def self.quoted_table_names # :nodoc:
+        if config_or_deprecated_connection.is_a?(Hash)
+          @raw_connection = nil
+          @config = config_or_deprecated_connection.symbolize_keys
+          @logger = ActiveRecord::Base.logger
```
Are we saying adapters can't set their own logger now? Do any adapters set their own logger? I think that'd be odd but not impossible.

matthewd replied:
My plan is to ultimately pull the connection's base model from the handler (or pass it in the supplied config hash?), and then use that object's logger -- so you can't directly override the adapter's logger during initialize (which hardly anyone calls anyway -- note that `mysql2_connection` and friends are already just passing `logger`), but there is some control beyond the hard-coded AR::Base.

eileencodes commented on activerecord/lib/active_record/connection_adapters/mysql2_adapter.rb:
```diff
@@ -120,12 +115,12 @@ def quote_string(string)
       def reconnect!(restore_transactions: false)
         @lock.synchronize do
-          @raw_connection.close
+          @raw_connection&.close
```
I don't like that now we have to keep checking for `nil` everywhere. I'm concerned we'll forget to do this in the future for a new method.

matthewd replied:
I think it's fairly fine... we have plenty of potentially-nil ivars in various places. The main mitigating factor here is that we should rarely add new methods that use the variable directly: outside of connection management (dis/re-/connect), nearly everything else should use `with_raw_connection`, `valid_raw_connection`, or `any_raw_connection` (in that preference order).

The other side is just that there isn't much of an alternative: while abstract adapter defines that `@raw_connection` exists, is nil when not connected, and truthy when connected... the rest of the API is adapter-specific, so it's impossible for us to use a null-object pattern, say.

eileencodes commented on activerecord/test/cases/adapters/mysql2/connection_test.rb:
```diff
@@ -55,21 +55,16 @@ def test_successful_reconnection_after_timeout_with_verify
 
-    assert_raise(ActiveRecord::ConnectionNotEstablished) do
-      @connection.quote("string")
-    end
```
Could this potentially be an undesirable change in behavior for applications?

matthewd replied:
_Theoretically_, yes -- I believe this is the one observable difference in existing API behaviour, outside of timing.

I did consider separately tracking a concept of explicit disconnection, and having that persist until an equally-explicit reconnect... but I just can't come up with a reasonable use case we're breaking to justify the extra code / state. The obvious use for `disconnect!` would be that you e.g. want to restart the DB server, so need to break your existing session... but we still do that; we just reconnect on demand instead of leaving the adapter in a semi-permanent state of brokenness.

matthewd committed:
Ensure a lost connection upon COMMIT fails "cleanly"

There's no point attempting to rollback in this case; we can instead
just invalidate the now-lost transaction.

Similarly, even though we can't immediately reconnect for a connection
failure mid-transaction, we can and should drop and prior verification
of the connection: the next out-of-transaction query attempt needs to
fix it, not assume it's working.

matthewd committed:
Configure legacy-API-supplied connection before first use

matthewd committed:
Non-retryable exceptions also break connection verification

matthewd committed:
We're still inside the transaction until the ROLLBACK has run

This previously made no difference, but it now saves us reconnecting to
run a ROLLBACK: because we're still inside a transaction at that moment,
we won't be #restorable?.

matthewd committed:
Drop default retry count to 1

This is enough to enable basic recovery for everyone by default, while
leaving more time-consuming repeated attempts for those who configure
them.

matthewd committed:
Be a bit clearer in the new-adapter-API argument error

matthewd committed:
Don't expect rollback to cause a reconnect

eileencodes commented:
Ran this against our app like we talked about and I mentioned some failures. Digging in a bit there's at least one issue here https://github.com/Shopify/semian/blob/ee8b77f1baa17ef71f543b406deb76e956597db3/lib/semian/rails.rb#L10 where we expect there to always be a connection. Before `@raw_connection` was never `nil`, now it can be `nil`. This happens in the middle of the test run so I'm not sure this is intentional and there may be a bug. I can dig in more tomorrow.
